echo: Convert module to TypeScript.

LaPulgaaa opened the pull request:
<!-- Describe your pull request here.-->

Fixes: <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

LaPulgaaa commented:
@zulipbot add "buddy review"

LaPulgaaa commented:
@zulipbot add "mentor review"

LaPulgaaa commented on web/src/state_data.ts:
```diff
@@ -189,7 +189,7 @@ export const onboarding_step_schema = one_time_notice_schema;
 const current_user_schema = z.object({
     avatar_source: z.string(),
-    avatar_url: NOT_TYPED_YET,
+    avatar_url: z.string().nullable(),
```
https://zulip.com/api/get-events#message

Lalit3716 commented on web/src/echo.ts:
```diff
@@ -367,14 +460,23 @@ export function update_message_lists({old_id, new_id}) {
 }
 
-export function process_from_server(messages) {
+export function process_from_server(messages: ServerMessage[]): ServerMessage[] {
```
Let's finally write a zod schema to parse the incoming message object from the server here. Ideally, we should parse it as soon as we get it from our events API which happens in `server_events.js` inside `get_events_success` method.
After writing the schema we can then infer the `ServerMessage` type from it.

LaPulgaaa replied:
I agree. Added as prep commit 

- https://github.com/zulip/zulip/pull/30702/commits/d82fadb7f6949ffc512951a0d93dfb40818e055b

Lalit3716 commented on web/src/echo.ts:
```diff
@@ -367,14 +460,23 @@ export function update_message_lists({old_id, new_id}) {
 
-export function process_from_server(messages) {
+export function process_from_server(messages: ServerMessage[]): ServerMessage[] {
     const msgs_to_rerender = [];
     const non_echo_messages = [];
```
These two arrays are implicitly of type `any[]` we should type them with `Message[]`.

LaPulgaaa replied:
Done.

LaPulgaaa commented on web/src/echo.ts:
```diff
@@ -19,17 +22,84 @@ import {current_user} from "./state_data";
+        | (PrivateMessageObject & {display_recipient?: DisplayRecipientUser[]})
+    );
+
+type Data = z.output<typeof send_message_api_response_schema>;
```
Sure!

timabbott replied:
Looks like you missed this suggestion, I'll tweak.

afeefuddin commented:
there are some linter errors, that need to be fixed.

LaPulgaaa commented:
The linter error is very wierd. 
```
Run source tools/ci/activate-venv
Using /srv/zulip-py3-venv
ts        | avoid subject in JS code at web/src/echo.ts line 115:
ts        |                 subject: z.string(),
ts        | avoid subject in JS code at web/src/echo.ts line 126:
ts        |                 subject: z.literal(""),
Error: Process completed with exit code 1.
```
This is suggesting me to remove `subject` in "JS code" . But `subject` is one of the fields in the `message` recieved from the server (and not a keyword or anything else **here**).

LaPulgaaa commented:
This is ready for another look @Lalit3716

afeefuddin commented on web/src/echo.js:
```diff
@@ -23,6 +24,104 @@ import * as util from "./util";
+
+export const server_message_schema = z
+    .object({
+        avatar: z.string().nullish(),
```
This should be `avatar_url`.

LaPulgaaa replied:
Fixed it.

LaPulgaaa commented:
@afeefuddin ready for your another review.

LaPulgaaa commented:
Although we are removing uses of `unknown_local_echo_user`, I did not remove https://github.com/zulip/zulip/pull/30702/commits/b25f2a60a77e5f6bc138f2dae4793d73d7b28b6b. The reason is that the rest of the frontend codebase still needs it until we have removed all the instances of `unknown_local_echo_user` -- I thought that might be a shaky refactor.

LaPulgaaa commented on web/tests/server_events.test.js:
```diff
@@ -51,14 +51,47 @@ const message_events = mock_esm("../src/message_events", {
 
 const server_events = zrequire("server_events");
 
+const message = {
```
The future plan is to use a `make_message` function to generate test data as we discussed.

LaPulgaaa commented on web/src/echo.js:
```diff
@@ -119,24 +119,6 @@ export function build_display_recipient(message) {
     const display_recipient = emails.map((email) => {
         email = email.trim();
         const person = people.get_by_email(email);
-        if (person === undefined) {
```
- https://github.com/zulip/zulip/pull/30509#discussion_r1651304420

afeefuddin commented on web/src/echo.ts:
```diff
@@ -119,24 +212,8 @@ export function build_display_recipient(message) {
     const display_recipient = emails.map((email) => {
         email = email.trim();
         const person = people.get_by_email(email);
-        if (person === undefined) {
-            // For unknown users, we return a skeleton object.
-            //
-            // This allows us to support zephyr mirroring situations
-            // where the server might dynamically create users in
-            // response to messages being sent to their email address.
-            //
-            // TODO: It might be cleaner for the web app for such
-            // dynamic user creation to happen inside a separate API
-            // call when the pill is constructed, and then enforcing
-            // the requirement that we have an actual user object in
-            // `people.js` when sending messages.
-            return {
-                email,
-                full_name: email,
-                unknown_local_echo_user: true,
-            };
-        }
```
If you are removing this then type `DisplayRecipentUser` will never contain `unknown_local_echo_user`. In that case, we should drop that completely. It is just used in one place.

LaPulgaaa replied:
Sure. https://github.com/zulip/zulip/pull/30702/commits/738b93517d86566f3bead2aef7e98ac501188eaa drops all instances of `unknown_local_echo_user` from frontend.

afeefuddin commented on web/src/echo.ts:
```diff
@@ -376,14 +461,23 @@ export function update_message_lists({old_id, new_id}) {
         const local_id = message.local_id;
+
+        if (local_id === undefined) {
+            // For messages that weren't locally echoed, we go through
+            // the "main" codepath that doesn't have to id reconciliation.
+            // We simply return non-echo messages to our caller.
```
Should we add same comment twice?

LaPulgaaa replied:
I believe the comments are "private" to their enclosing `if` blocks, hence I put them their twice.

timabbott replied:
I ended up reorganizing the comments a bit for more clarity.

afeefuddin commented on web/src/echo.ts:
```diff
@@ -10,27 +12,103 @@ import * as markdown from "./markdown";
 
-const waiting_for_id = new Map();
-let waiting_for_ack = new Map();
+const server_message_schema_with_flags = server_message_schema.and(
+    z.object({
+        local_id: z.string().optional(),
+        flags: z.array(z.string()),
+    }),
+);
+
+type ServerMessage = z.output<typeof server_message_schema_with_flags>;
```
I think we can use `type RawMessage` instead of creating a new type `ServerMessage`. 
I used a similar schema in this commit ae05786a24c8ffd0fa6ac38ba0fc4b3a2bdd46ca to infer the type `RawMessage`.

I believe we should avoid making too many new types since it will create a problem migrating other files due to these different types with barely 1 key different from the other.

LaPulgaaa replied:
I mostly agree with your point except for few things
- I think messages recieved from server should have an independent type as `raw_message` and `server_message` differ in couple of fields.
- Also even if we do not create a new type and use `Omit<RawMessage, "foo field"> & {bar fields}`, I think there are few incomplete type fields in `RawMessage` that have to be fixed first.

afeefuddin commented on web/src/echo.ts:
```diff
@@ -162,54 +240,59 @@ export function build_display_recipient(message) {
+        id: local_id_float,
+        resend: false,
+        is_me_message: false,
+        topic_links: topic !== undefined ? markdown.get_topic_links(topic) : [],
```
I think the topic will never be `undefined` here instead it will be an empty string in case we are sending a private message. Hence `topic !== undefined` will always be true. I guess the intention here is to check if the `topic.length >0`.

LaPulgaaa replied:
Yeah the actually should have been for falsy (empty string). Thanks for catching that. It was a potential bug fixed in https://github.com/zulip/zulip/pull/30702/commits/7570fbddca68e27a8654a40eff39deb1aa76206e

LaPulgaaa commented:
@afeefuddin I have addressed all the review comments, but this module is not ready for migration yet, as 
- https://github.com/zulip/zulip/pull/30813
introduces a new module(`message_events_util.js`) dependency on `echo.js`. So might need to handle this first. Probably migrating `message_events_util.js` before `echo.js`

LaPulgaaa commented:
@afeefuddin rebasing this since 
- https://github.com/zulip/zulip/pull/31145
is merged now !!

timabbott commented on web/src/server_message.ts:
```diff
@@ -31,7 +31,7 @@ const message_reaction_schema = z.array(
             email: z.string(),
             full_name: z.string(),
-            is_mirror_dummy: z.boolean().nullish(),
+            is_mirror_dummy: z.boolean(),
```
So this is the schema, but maybe we should be ignoring this block?

https://zulip.com/api/get-messages notes this as deprecated, and we could just stop parsing it.

timabbott replied:
Posted in API design, but it looks like we might be removing this in the next month or two. So probably useful refactoring to do to stop parsing it, but not urgent.

LaPulgaaa replied:
Yeah, it is OK to remove this field  **now**. Even if the server keeps sending this (until we remove this completely) `zod` will ignore the field.  As a result no downstream code will have the field (which is what we want).

LaPulgaaa replied:
Done in https://github.com/zulip/zulip/pull/30702/commits/c18348c9c6cc24a9e0d40ed97b8d811264068047 .

LaPulgaaa commented:
Ready for another review @afeefuddin

afeefuddin commented on web/src/echo.ts:
```diff
@@ -153,6 +221,7 @@ export function build_display_recipient(message) {
             id: person.user_id,
             email: person.email,
             full_name: person.full_name,
+            is_mirror_dummy: false,
```
Can we drop `is_mirror_dummy` from the type? I don't see it being used anywhere.
Am I missing something?

LaPulgaaa replied:
Done in https://github.com/zulip/zulip/pull/30702/commits/92a40d80a83c3ac8208d50218d3f23ceac190b8d

timabbott replied:
We do have upcoming features that might use it, so it is probably better not to drop it.

afeefuddin commented:
Looks good to me!!

LaPulgaaa commented:
@zulipbot remove "buddy review"

LaPulgaaa commented:
@zulipbot add "integration review"

timabbott commented on web/src/echo.ts:
```diff
@@ -21,17 +24,84 @@ import {current_user} from "./state_data";
+    reply_to: string;
+    private_message_recipient: string;
+    to_user_ids: string | undefined;
+};
```
Hmm, the server API I think only expects `to_user_ids`.

timabbott commented on web/src/echo.ts:
```diff
@@ -21,17 +24,84 @@ import {current_user} from "./state_data";
+    alerted: boolean;
+    mentioned: boolean;
+    mentioned_me_directly: boolean;
+}>;
```
Renamed this to `LocalEditRequest`, since it's for locally echoing the EDIT of an existing message.

timabbott commented on web/src/echo.ts:
```diff
@@ -61,28 +131,43 @@ function show_message_failed(message_id, failed_msg) {
-function failed_message_success(message_id) {
-    message_store.get(message_id).failed_request = false;
+function failed_message_success(message_id: number): void {
+    message_store.get(message_id)!.failed_request = false;
```
This one could could be reworked in a follow-up to be passed the actual message object rather than looking it up, to remove this type assertion.

LaPulgaaa replied:
Fetching `message` inside a closure(`on_success` callback) that might have an old state of message_store( the state at the time of its creation) and then passing it to `failed_message_success` might not be correct. I tried to do it and the message fetched inside the closure is `undefined`. Most likely due to the old_state that it had captured during its creation.

timabbott commented on web/src/echo.ts:
```diff
@@ -61,28 +131,43 @@ function show_message_failed(message_id, failed_msg) {
+        ) => void;
+    },
+): void {
+    message.content = message.raw_content!;
```
This is pretty weird code; I wonder if there's a nicer way to do the types for it.

timabbott commented on web/src/echo.ts:
```diff
@@ -61,28 +131,43 @@ function show_message_failed(message_id, failed_msg) {
+        on_send_message_success: (request: Message, data: Data) => void;
+        send_message: (
+            request: Message,
+            on_success: (data: unknown) => void,
```
This should be `raw_data: unknown` for consistency.

timabbott replied:
Changed.

timabbott commented on web/src/echo.ts:
```diff
@@ -292,9 +392,9 @@ export function edit_locally(message, request) {
-            message.alerted = request.alerted;
+            message.mentioned = request.mentioned ?? false;
+            message.mentioned_me_directly = request.mentioned_me_directly ?? false;
+            message.alerted = request.alerted ?? false;
```
Hmm, are these not always available in these requests?

LaPulgaaa replied:
Nope.

timabbott commented on web/src/message_store.ts:
```diff
@@ -117,6 +118,10 @@ export type Message = (
+    failed_request?: boolean;
+    show_slow_send_spinner?: boolean;
+    resend?: boolean;
+    local_id?: string;
```
I added some comments around this block of local-echo related state.

timabbott commented:
I fixed the comments above that I felt blocking for merging; the rest seem like things we can investigate and address in localized follow-up PRs, since they seem unlikely to contaminate typing for other modules. 

Merged, thanks for all the effort on this @LaPulgaaa and @afeefuddin! This is a really big one.
